File: backend/app/models/robot.py
from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class RobotStatus(StrEnum):
    IDLE = "IDLE"
    CLEANING = "CLEANING"
    CHARGING = "CHARGING"
    RETURNING_TO_DOCK = "RETURNING_TO_DOCK"
    LOW_BATTERY = "LOW_BATTERY"
    UPDATING = "UPDATING"
    ERROR = "ERROR"


class Position(BaseModel):
    x: float
    y: float
    room: str


class Robot(BaseModel):
    robot_id: str
    name: str = ""
    status: RobotStatus = RobotStatus.IDLE
    battery_level: float = 100.0
    position: Position = Field(default_factory=lambda: Position(x=0.5, y=0.5, room="charging_dock"))
    speed: float = 0.0
    firmware_version: str = "1.0.0"
    last_seen: str = ""
    error_code: str | None = None


class RobotCommand(BaseModel):
    command: str
    params: dict[str, Any] = Field(default_factory=dict)


class CommandResponse(BaseModel):
    accepted: bool
    message: str = ""

File: frontend/src/types/robot.ts
export type RobotStatus =
  | 'IDLE'
  | 'CLEANING'
  | 'CHARGING'
  | 'RETURNING_TO_DOCK'
  | 'LOW_BATTERY'
  | 'UPDATING'
  | 'ERROR';

export interface Position {
  x: number;
  y: number;
  room: string;
}

export interface Robot {
  robot_id: string;
  name: string;
  status: RobotStatus;
  battery_level: number;
  position: Position;
  speed: number;
  firmware_version: string;
  last_seen: string;
  error_code?: string | null;
}

export interface TelemetryPoint {
  timestamp: string;
  battery_level: number;
  speed: number;
  status: string;
}

export interface TelemetryHistory {
  robot_id: string;
  points: TelemetryPoint[];
}

export interface RobotCommand {
  command: string;
  params?: Record<string, unknown>;
}

export interface Schedule {
  schedule_id: string;
  robot_id: string;
  room_id: string;
  cron_expression: string;
  description: string;
  enabled: boolean;
}

export interface ScheduleCreate {
  robot_id: string;
  room_id: string;
  cron_expression: string;
  description?: string;
}

export interface OtaJob {
  job_id: string;
  robot_id: string;
  status: string;
  firmware_version: string;
  new_speed: number;
  progress: number;
  started_at: string;
  completed_at: string;
}

export interface OtaJobCreate {
  robot_ids: string[];
  new_speed: number;
  version: string;
  description?: string;
}

// ─── WebSocket ─────────────────────────────────────────

export type WsMessage =
  | { type: 'initial_state'; robots: Robot[] }
  | { type: 'robot_update'; robot: Robot }
  | { type: 'alert'; robot_id: string; message: string };
